phoenix/src/recovery.rs

```diff
@@ -0,0 +1,222 @@
+use heapless::HistoryBuffer;
+use libm::powf;
+// --- Boom Boom Setup ---
+/*
+   MAIN_ARM/TEST = PD6
+   MAIN_FIRE = PD5
+   MAIN_ARM/TEST_B = PD14
+   MAIN_FIRE_B = PD13
+   DROGUE_ARM/TEST = PC11
+   DROGUE_FIRE = PC12
+   DROGUE_ARM/TEST_B = PD2
+   DROGUE_FIRE_B = PD1
+   MAIN_MCU_EMATCH_SENSE = PA2
+   MAIN_MCU_EMATCH_SENSE_B = PB0
+   DROUGE_MCU_EMATCH_SENSE = PA3
+   DROGUE_MCU_EMATCH_SENSE_B = PC5
+*/
+
+struct Arming {
+    main: Output<'static>,
+    drogue: Output<'static>,
+    main_b: Output<'static>,
+    drogue_b: Output<'static>,
+}
+
+struct Fire {
+    main: Output<'static>,
+    drogue: Output<'static>,
+    main_b: Output<'static>,
+    drogue_b: Output<'static>,
+}
+
+struct Sensing {
+    adc: embassy_stm32::peripherals::ADC1,
+    main: PA2,
+    main_b: PB0,
+    drogue: PA3,
+    drogue_b: PC5,
+}
+
+pub struct RecoveryManager {
+    ejection_enable: Input<'static>,
+    sensing: Sensing,
+    arming: Arming,
+    fire: Fire,
+}
+
+impl RecoveryManager {
+    pub fn new(
+        main_arm: PD6,
+        main_arm_b: PD14,
+        drogue_arm: PC11,
+        drogue_arm_b: PD2,
+        main_fire: PD5,
+        main_fire_b: PD13,
+        drogue_fire: PC12,
+        drogue_fire_b: PD1,
+        main_sense: PA2,
+        main_b_sense: PB0,
+        drogue_sense: PA3,
+        drogue_sense_b: PC5,
+        adc: ADC1,
+        ejection_enable: PC1,
+    ) -> Self {
+        RecoveryManager {
+            ejection_enable: Input::new(ejection_enable, Pull::Down),
+            sensing: Sensing {
+                adc,
+                main: main_sense,
+                main_b: main_b_sense,
+                drogue: drogue_sense,
+                drogue_b: drogue_sense_b,
+            },
+            arming: Arming {
+                main: Output::new(main_arm, Level::Low, Speed::Low),
+                main_b: Output::new(main_arm_b, Level::Low, Speed::Low),
+                drogue: Output::new(drogue_arm, Level::Low, Speed::Low),
+                drogue_b: Output::new(drogue_arm_b, Level::Low, Speed::Low),
+            },
+            fire: Fire {
+                main: Output::new(main_fire, Level::Low, Speed::Low),
+                main_b: Output::new(main_fire_b, Level::Low, Speed::Low),
+                drogue: Output::new(drogue_fire, Level::Low, Speed::Low),
+                drogue_b: Output::new(drogue_fire_b, Level::Low, Speed::Low),
+            },
+        }
+    }
+
+    pub fn arm(&mut self) {
+        if self.ejection_enable.is_high() {
+            info!("arm ejection enabled");
+            self.arming.main.set_high();
+            self.arming.main_b.set_high();
+
+            self.arming.drogue.set_high();
+            self.arming.main_b.set_high();
+        }
+    }
+
+    pub fn disarm(&mut self) {
+        info!("arm ejection disabled");
+        self.arming.main.set_low();
+        self.arming.main_b.set_low();
+
+        self.arming.drogue.set_low();
+        self.arming.main_b.set_low();
+    }
```

Comment:
Any reason to call `self.arming.main_b.set_low();` and ` self.arming.main_b.set_high();` in rapid succession? like that?